### PR TITLE
[Feature] vLLM CLI for serving and querying OpenAI compatible server

### PR DESCRIPTION
**Pull request by EthanqX**
FILL IN THE PR DESCRIPTION HERE

FIX #4167 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Thanks for picking up where @simon-mo left off! Please remember to address the comments that originated in #4167.

File: vllm/scripts.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
+# The CLI entrypoint to vLLM.
+import argparse
+import requests
```
Comment (simon-mo) [inline vllm/scripts.py:3]: Let's just `import openai` so you are using this https://github.com/openai/openai-python

File: vllm/scripts.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    serve_parser = make_arg_parser(serve_parser)
+    serve_parser.set_defaults(func=run_server)
+    
+    complete_parser = subparsers.add_parser(
+        "complete",
+        help="Generate text completions based on the given prompt via the running API server",
+        usage="vllm complete [options]")
+    complete_parser.add_argument("--url",
+                                 type=str,
+                                 default=None,
```
Comment (simon-mo) [inline vllm/scripts.py:36]: default to "http://localhost:8000" so user can put in `https://api.openai.com` as well

Comment (simon-mo) [reply vllm/scripts.py:36]: Yes let's also add an option for api-key

File: vllm/scripts.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                                 help="url of the running OpenAI-Compatible RESTful API server")
+    complete_parser.add_argument("--complete-model-name",
+                                type=str, default=None,
+                                help="the model name used in prompt completion")
```
Comment (simon-mo) [inline vllm/scripts.py:40]:
```suggestion
                                help="the model name used in prompt completion, default to the first model in list models API call.")
```

File: vllm/scripts.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    chat_parser.add_argument("--url",
+                                 type=str,
+                                 default=None,
+                                 help="url of the running OpenAI-Compatible RESTful API server")
+    chat_parser.add_argument("--chat-model-name",
+                                type=str, default=None,
+                                help="the model name used in chat completions")
+    chat_parser.add_argument("--generation-prompt",
+                             type=str,
+                             help="the generation prompt to be added to the chat template")
```
Comment (simon-mo) [inline vllm/scripts.py:57]:
```suggestion
```

Comment (simon-mo) [reply vllm/scripts.py:57]: However, let's add an option for `--system-prompt` so for method that support system prompt user can put in one.

File: vllm/scripts.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        complete(args)
+    elif args.command == "chat":
+        chat(args)
+
+
+def complete(args: argparse.Namespace) -> None:
+    while True:
+        input_prompt = input("Please enter prompt to complete:")
+
+        base_url = getattr(args, "url", f"http://{args.host}:{args.port}")
```
Comment (simon-mo) [inline vllm/scripts.py:78]: to be fixed by the url field.

File: vllm/scripts.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        
+        try:
+            completion = _post_response_dict(create_completion_url,
+                                            create_completion_headers,
+                                            completion_request_data)
+            choice = completion.get("choices", [])[0].get("text", "No completion found.")
+            print(f"Response Content: {choice}")
+        except Exception as err:
+            print(f"Error occurred in prompt completion: {err}")
+            return
```
Comment (simon-mo) [inline vllm/scripts.py:103]: Let's use OpenAI Python SDK

File: vllm/scripts.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+
+def chat(args: argparse.Namespace) -> None:
+    conversation = []
```
Comment (simon-mo) [inline vllm/scripts.py:107]: TODO: handle system prompt if supplied.

File: vllm/scripts.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    while True:
+        input_message = input("Please enter a message for the chat model:")
```
Comment (simon-mo) [inline vllm/scripts.py:109]:
```suggestion
        input_message = input("Please enter a message for the chat model:\n>")
```

File: vllm/scripts.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        message = {
+            "role": "user",
+            "content": input_message
+        }
+        if args.generation_prompt:
+            message["generation prompt"] = args.generation_prompt
```
Comment (simon-mo) [inline vllm/scripts.py:115]:
```suggestion
```

File: vllm/scripts.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+                chat_completion.get("content", {})
+                               .get("choices", [])[0]
+                               .get("message", {})
+                               .get("content", "No chat completion found.")
+            )
+            print(f"Response Content: {choice}")
+        except Exception as err:
+            print(f"Error occurred in chat completion: {err}")
+            return
+
```
Comment (simon-mo) [inline vllm/scripts.py:149]: TODO: append the response here as well.

File: vllm/scripts.py [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    try:
+        response = requests.get(url)
+        response.raise_for_status()
+        return response.json()
+    except requests.exceptions.HTTPError as http_err:
+        raise Exception(f"HTTP error occurred: {http_err}")
+    except requests.exceptions.JSONDecodeError as json_err:
+        raise Exception(f"Invalid JSON response: {json_err}")
+    except Exception as err:
+        raise Exception(f"Error occurred: {err}")
```
Comment (simon-mo) [inline vllm/scripts.py:173]: you don't need these for Python SDK

**Review by simon-mo — CHANGES_REQUESTED** [review]

**Comment by simon-mo** [comment]
@rkooo567 I pushed some changes. Please take another look!

**Comment by simon-mo** [comment]
I also reverted the doc change. We should test this out for another commit

**Review by simon-mo — APPROVED** [review]

File: vllm/scripts.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+    print(f"Using model: {model_name}")
+
+    if args.command == "complete":
+        complete(model_name, openai_client)
+    elif args.command == "chat":
+        chat(args.system_prompt, model_name, openai_client)
+
+
+def complete(model_name: str, client: OpenAI) -> None:
```
Comment (rkooo567) [inline vllm/scripts.py:55]: QQ: do we support setting some sampling params?

**Review by rkooo567 — COMMENTED** [review]
Looks pretty good to me. But I think we need to add some basic unit tests? (at least basic sanity check for serve -> chat and serve -> complete?)

**Review by rkooo567 — APPROVED** [review]

**Comment by DarkLight1337** [comment]
You should update the docs to replace any references to the old way of launching API server.

Please also apply the change mentioned [here](https://github.com/vllm-project/vllm/pull/4167#discussion_r1595291962).

File: vllm/scripts.py (hunk 12 above)
Comment (simon-mo) [reply vllm/scripts.py:55]: Yes. https://docs.vllm.ai/en/stable/serving/openai_compatible_server.html#extra-parameters

File: vllm/entrypoints/openai/api_server.py [hunk 13]
```diff
@@ -219,3 +231,13 @@ async def authentication(request: Request, call_next):
                 ssl_certfile=args.ssl_certfile,
                 ssl_ca_certs=args.ssl_ca_certs,
                 ssl_cert_reqs=args.ssl_cert_reqs)
+
+
+if __name__ == "__main__":
+    # NOTE(simon):
+    # This section should be in sync with vllm/scripts.py for CLI entrypoints.
```
[the remark this answers is not in the document]
Comment (simon-mo) [reply vllm/entrypoints/openai/api_server.py:238]: In sync in their usage of `make_arg_parser`

File: vllm/entrypoints/openai/api_server.py [hunk 14]
```diff
@@ -174,6 +174,12 @@ async def authentication(request: Request, call_next):
             raise ValueError(f"Invalid middleware {middleware}. "
                              f"Must be a function or a class.")
 
+    return app
+
+
+def run_server(args):
```
Comment (prashantgupta24) [inline vllm/entrypoints/openai/api_server.py:180]: Is it possible to make `engine` as an optional arg to this function?
```suggestion
def run_server(args, llm_engine=None)
```
This can help external applications reuse the llm engine and attach other API interfaces (like grpc) to the same llm engine. To be used with the other suggestion below

File: docs/source/serving/openai_compatible_server.md [hunk 15]
```diff
@@ -109,7 +109,7 @@ directory [here](https://github.com/vllm-project/vllm/tree/main/examples/)
 
 ```{argparse}
 :module: vllm.entrypoints.openai.cli_args
-:func: make_arg_parser
+:func: create_parser_for_docs
 :prog: -m vllm.entrypoints.openai.api_server
```
Comment (DarkLight1337) [inline docs/source/serving/openai_compatible_server.md:113]: Can we change this to `vllm serve`?

Comment (DarkLight1337) [inline docs/source/serving/openai_compatible_server.md:113]: Can we change this to `vllm serve`?

File: vllm/entrypoints/openai/api_server.py (hunk 14 above)
Comment (dtrifiro) [reply vllm/entrypoints/openai/api_server.py:180]: +1, this would be useful.

**Review by rafvasq — COMMENTED** [review]
I'm just curious if documentation should be updated to use this CLI as well, for example:

- [Quickstart](https://github.com/vllm-project/vllm/blob/main/docs/source/getting_started/quickstart.rst#openai-compatible-server)
- [Adding Model](https://github.com/vllm-project/vllm/blob/main/docs/source/models/adding_model.rst?plain=1#L113)
- [Lora](https://github.com/vllm-project/vllm/blob/main/docs/source/models/lora.rst?plain=1#L61)

**Comment by youkaichao** [comment]
@EthanqX hi, this is very cool. any blocker now?

**Comment by youkaichao** [comment]
LGTM, we can merge this if the full test passes 👍

**Comment by EthanqX** [comment]
@youkaichao Ready to merge.

File: docs/source/serving/openai_compatible_server.md (hunk 15 above)
Comment (EthanqX) [reply docs/source/serving/openai_compatible_server.md:113]: Documentation remains unchanged for this release.

File: vllm/scripts.py [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+        help="Generate chat completions via the running API server",
+        usage="vllm chat [options]")
+    _add_query_options(chat_parser)
+    chat_parser.add_argument(
+        "--system-prompt",
+        type=str,
+        default=None,
+        help=("The system prompt to be added to the chat template, "
+              "used for models that support system prompts."))
+    chat_parser.set_defaults(dispatch_function=interactive_cli, command="chat")
```
Comment (ywang96) [inline vllm/scripts.py:143]: I would add some documentation on how to use these (can be in a separate PR) if we're planning to release this feature.

File: vllm/scripts.py (hunk 12 above)
Comment (ywang96) [reply vllm/scripts.py:52]: I think Sang's question is whether or not we should support setting sampling params through `complete` and `chat`; for now it's using all default parameters.

```
        completion = client.completions.create(model=model_name,
                                               prompt=input_prompt)
```

**Review by ywang96 — COMMENTED** [review]
I know this is merged already but I left two comments!

File: vllm/scripts.py (hunk 12 above)
Comment (EthanqX) [reply vllm/scripts.py:52]: Indeed, it would require checking field by field whether any sampling params are provided to override the default value.
